readme.txt
#       Copyright 2010 Mephiston <meph.snake@gmail.com>
#
#       This program is free software; you can redistribute it and/or modify
#       it under the terms of the GNU General Public License as published by
#       the Free Software Foundation; either version 3.0 of the License, or
#       (at your option) any later version.
#
#       This program is distributed in the hope that it will be useful,
#       but WITHOUT ANY WARRANTY; without even the implied warranty of
#       MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#       GNU General Public License for more details.
#
#       You should have received a copy of the GNU General Public License
#       along with this program; if not, get a copy on http://www.gnu.org/licenses/gpl.txt

#   ---------------- Generic utilities ---------------- #
#   A usefull an fun bot for irc using python-irclib, and ircbot of Joel Rosdahl <joel@rosdahl.net> as base.


python-irclib/Changelog
2008-09-11  Keltus <keltus@users.sourceforge.net>
    * ircbot.py: Reverted a leftover testing fragment that broke ircbot.py

2008-08-29  Keltus <keltus@users.sourceforge.net>
    * irclib.py: Reverted changes in irclib that don't really belong in the core lib
    * irclib.py: Add support for ssl
    * irclib.py: Add support for ipv6
	* python-irclib now requires python 2.3 and above (for ssl and ipv6)

2006-03-09  Keltus <keltus@users.sourceforge.net>
	* README: Minor update

2005-12-24  Keltus <keltus@users.sourceforge.net>
	* Released version 0.4.6.

	* irclib.py (VERSION):
	* python-irclib.spec.in:
	Preparations for version 0.4.6.

2005-12-23  Keltus <keltus@users.sourceforge.net>
	* dccsend:
	* dccreceive:
	* irclib.py:
	* ircbot.py:
	* irccat:
	* irccat2:
	* servermap:
	* testbot.py:
	Code modernization - String methods used instead of deprecated
	string functions, keyword 'in' used for membership testing instead
	of 'has_key' method, etc.

2005-12-06  Keltus <keltus@users.sourceforge.net>
	* irclib.py (ServerConnection.process_data): Reversed fix from
	2005-05-28. This is strange because there was a bug before and
	now it's gone. Either python changed something, or the IRC
	networks changed something. Confirmed by peter.

2005-11-03  Keltus <keltus@users.sourceforge.net>
	* irclib.py (numeric_events): Renamed numeric code 332 from topic
	to currenttopic (the message when "/topic <chan>" is sent), so it
	doesn't collide with TOPIC (the message when the topic is set).

2005-08-27  Keltus <keltus@users.sourceforge.net>
	* irclib.py (ServerConnection.disconnect): Fixed infinitely
	recursive calls when disconnecting with a failed connection. Bug
	reported by Erik Max Francis.

2005-08-18  Keltus <keltus@users.sourceforge.net>
	* irclib.py: Made ServerConnection.disconnect more consistant and
	changed some functions to use it instead of quit. Previously,
	disconnect would ignore the quit message, but now it sends a quit
	message and disconnect. Suggestion by Erik Max Francis.
	* ircbot.py: Changed to use ServerConnection.disconnect instead of
	ServerConnection.quit as well.

2005-05-28  Keltus <keltus@users.sourceforge.net>
	* irclib.py (ServerConnection.process_data): Fixed quit arguments
	to return a list rather than a list of a list. Patch from peter.

2005-05-18  Keltus <keltus@users.sourceforge.net>
	* Released version 0.4.5.

	* irclib.py (ServerConnection.__init__): Added self.socket = None
	to be able to process events when ServerConnection is not
	connected to a socket. Patch from alst.

	* irclib.py (VERSION):
	* python-irclib.spec.in:
	Preparations for version 0.4.5.

2005-04-26  Keltus <keltus@users.sourceforge.net>
	* irclib.py (IRC.__doc__): Corrected server.process_forever() to
	irc.process_forever(). Suggestion by olecom.

2005-04-17  Keltus <keltus@users.sourceforge.net>
	* irclib.py (ServerConnection.process_data): Moved event
	translation code.
	* irclib.py (ServerConnection): Reverted the 2005-01-28 change
	because it breaks jump_server().
	* irclib.py: minor comment changes

2005-04-03  Keltus <keltus@users.sourceforge.net>
	* irclib.py (protocol_events): Added "pong" and "invite" events.
	Patch from Adam Mikuta.
	* irclib.py (ServerConnection.part): Added message parameter.
	Patch from Adam Mikuta.

2005-02-23  Keltus <keltus@users.sourceforge.net>
	* Released version 0.4.4.

	* irclib.py (VERSION):
	* python-irclib.spec.in:
	Preparations for version 0.4.4.

2005-01-28  Keltus <keltus@users.sourceforge.net>
	* irclib.py: (ServerConnection): Moved
	self.irclibobj._remove_connection call from	close() to
	disconnect(). Patch from Alexey Nezhdanov.

2005-01-25  Keltus <keltus@users.sourceforge.net>
	* irclib.py (ServerConnection.connect): closes socket if a
	connection does not occur
	* irclib.py (ServerConnection.connect): "Changing server" ->
	"Changing servers" (more ubiquitous quit phrase)

2005-01-23  Keltus <keltus@users.sourceforge.net>

	* irclib.py: Removed depreciated apply functions. python-irclib is
	now compatible with Python 1.6 and above.
	* testbot.py: Removed redundant extra start() call

2005-01-20  Joel Rosdahl  <joel@rosdahl.net>

	* Released version 0.4.3.
	* Makefile: Removed more GNU make specific constructs.

2005-01-19  Joel Rosdahl  <joel@rosdahl.net>

	* Makefile: Don't require GNU make.

2005-01-19  Keltus  <keltus@users.sourceforge.net>

	* ircbot.py (IRCDict.__iter__): Added __iter__ method for IRCDict.

2005-01-17  Joel Rosdahl  <joel@rosdahl.net>

	* ircbot.py (IRCDict.__contains__): Added __contains__method for
	IRCDict. Patch from Keltus.
	(SingleServerIRCBot.on_ctcp): Corrected default decoding of CTCP
	DCC CHAT. Patch from Keltus.

	* irclib.py (VERSION):
	* python-irclib.spec.in:
	Preparations for version 0.4.3.

	* debian: Removed Debian package directory since python-irclib is
	in Debian now.

	* ircbot.py (SingleServerIRCBot._on_namreply): Improved comment
	about arguments to the function. Patch from Keltus.
	(Channel.has_allow_external_messages): Renamed from
	has_message_from_outside_protection. Patch from Keltus.

	* irclib.py (ServerConnection.quit): Added comment about how some
	IRC servers' treat QUIT messages. Patch from Keltus.

	* ircbot.py (SingleServerIRCBot.jump_server): Improved jump_server
	behaviour. Patch from Keltus.

2004-08-04  Joel Rosdahl  <joel@rosdahl.net>

	* irclib.py (ServerConnection.process_data): Added "bonus" action
	event that is triggered on CTCP ACTION messages.

2004-07-09  Joel Rosdahl  <joel@rosdahl.net>

	* Released version 0.4.2.

	* debian/rules: Remove built *.pyc files before making package.

	* irclib.py (DEBUG):
	* debian/changelog:
	* python-irclib.spec.in:
	Preparations for version 0.4.2.

	* irclib.py (ServerNotConnectedError): New exception.
	(ServerConnection.send_raw): Fix bug #922446, "Raise
	IllegalStateException in send_raw when disconnected".

2003-10-30  Joel Rosdahl  <joel@rosdahl.net>

	* Released version 0.4.1.


python-irclib/pyBoter64.py
#! /usr/bin/env python
#*.*coding:utf-8*.*
#
# Example program using ircbot.py.
#
# Joel Rosdahl <joel@rosdahl.net>

#       pyboter64.py written in python2.8
#       version 2.64
#       Copyright 2010 Mephiston <meph.snake@gmail.com>
#
#       This program is free software; you can redistribute it and/or modify
#       it under the terms of the GNU General Public License as published by
#       the Free Software Foundation; either version 3.0 of the License, or
#       (at your option) any later version.
#
#       This program is distributed in the hope that it will be useful,
#       but WITHOUT ANY WARRANTY; without even the implied warranty of
#       MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#       GNU General Public License for more details.
#
#       You should have received a copy of the GNU General Public License
#       along with this program; if not, get a copy on http://www.gnu.org/licenses/gpl.txt

#   ---------------- Generic utilities ---------------- #
#   A usefull an fun bot for irc using python-irclib, and ircbot of Joel Rosdahl <joel@rosdahl.net> as base.

"""A example bot.

This is an example bot that uses the SingleServerIRCBot class from
ircbot.py.  The bot enters a channel and listens for commands in
private messages and channel traffic.  Commands in channel messages
are given by prefixing the text by the bot name followed by a colon.
It also responds to DCC CHAT invitations and echos data sent in such
sessions.

The known commands are:

    stats -- Prints some channel information.

    disconnect -- Disconnect the bot.  The bot will try to reconnect
                  after 60 seconds.

    die -- Let the bot cease to exist.

    dcc -- Let the bot invite you to a DCC CHAT connection.
"""
#import ircbot
PDO_FLAG = True
try:
    from ircbot import SingleServerIRCBot
except:
    print "python-ircbot Not found."
try:
    from irclib import nm_to_n, nm_to_h, irc_lower, ip_numstr_to_quad, ip_quad_to_numstr
except:
    print "python-irclib Not found."
try:
    from httplib import HTTPConnection
except:
    print "python-httplib Not found."
try:
    import pdo
except:
    PDO_FLAG = False
    print "python-pdo Not found."
    
try:
    from BeautifulSoup import BeautifulSoup
except:
    print "python-beautifulsoup Not found"
    
from datetime import date, datetime
from time import time, asctime, localtime, strftime, sleep
from random import shuffle
from string import lowercase
from urllib import urlencode, urlopen
from sys import argv, exc_info
from traceback import format_exc
#import irclib


#~ VARIABLES
VERSION = '2.64'
DEBUG = 1
#~ if PDO_FLAG == True:
    #~ DB = pdo.connect("Module=MySQLdb;User=pyboter;Passwd=pyboter1.0;DB=pyboter;host=localhost")
    
class TestBot(SingleServerIRCBot):
    def __init__(self, channel, varNickname, varPassword, varServer, varPort=6667):
        SingleServerIRCBot.__init__(self, [(varServer, varPort)], varNickname, varNickname)
        self.channel = channel
        self.NickName = varNickname
        self.password = varPassword
        self.flood = {}
        self.previousNick = ''
        self.details = {
            "status": 0,
            "select_nick": "SELECT `perm` FROM `users` WHERE `nick`='%s' AND `password`='%s' AND (%s) LIMIT 1",
            "insert_nick": "INSERT IGNORE INTO `users` (`nick`,`password`,`perm`) values('%s', '%s', '%s');",
            "delete_nick": "DELETE FROM `%s` WHERE `nick`='%s';",
            "update_nick": "UPDATE `users` SET `password`='%s' WHERE `nick`='%s';",
            "quote": "SELECT `texts` FROM `quotes` WHERE `type`='%.1s' ORDER BY RAND() LIMIT 1,1;",
            "owners": ["mephiston", "meph"],
            "owned": False
        }
        self.log=''
        
        
    def sql_on(self, varChat, varEvent):
        global conn
        details["status"]=1
        db=pdo.connect("Module=MySQLdb;User=pyboter;Passwd=pyboter1.0;DB=pyboter;host=localhost")
        varChat.privmsg( self.channel, "La conexión con SQL esta apagada.")
    
    def sql_off(self, varChat, varEvent):
        details["status"]=0
        try:
            db.close()
            varChat.privmsg( self.channel, "La conexión con SQL esta apagada.")
        except:
            return False
            
    def show_credits(self, varChat, varEvent):
        actual_time = asctime(localtime())
        ch_active = varEvent.target()
        varChat.privmsg( ch_active, "%s" % actual_time)
        varChat.privmsg( ch_active, "Este bot está enteramente creado en Python y PDO.")
        varChat.privmsg( ch_active, "Un experimento para demostrar el potencial de Python.")
        varChat.privmsg( ch_active, "Creado por Daniel Ripol (Mephiston).")
        varChat.privmsg( ch_active, "En fecha: 19 de Setiembre de 2010.")
        varChat.privmsg( ch_active, "Agradecimientos a: Choms, vgvgf, galrunch, y Dahrkael.")
        self.log+=ch_active+" "+strftime("(%Y-%m-%d %H:%M:%S) ",localtime(time()))+self.NickName+"Este bot está enteramente creado en Python y PDO. Un experimento para demostrar el potencial de Python. Creado por Daniel Ripol (Mephiston). En fecha: 19 de Setiembre de 2010. Agradecimientos a: Choms, vgvgf, galrunch, y Dahrkael.\n"
        
    def show_version(self, varChat, varEvent):
        dcc = self.dcc_listen()
        nick = nm_to_n(varEvent.source())
        varChat.ctcp("VERSION", nick, "pyBoter 2.64 Python IRC %s %d" % (
            ip_quad_to_numstr(dcc.localaddress),
            dcc.localport))
            
    def show_image(self, varChat, varEvent):
        ch_active = varEvent.target()
        varSentence="".join(varEvent.arguments()[0].split(":", 1)).split()
        if len(varSentence) <2:
            mens="Por favor, especifica sobre que quieres buscar imágenes. Ejemplo: !imagen Python"
            varChat.privmsg( ch_active, mens)
            self.log+=ch_active+" "+strftime("(%Y-%m-%d %H:%M:%S) ",localtime(time()))+self.NickName+mens+"\n"
        else:
            for imagen in varSentence[1:]:
                soup = BeautifulSoup((urlopen("http://www.flickr.com/badge_code_v2.gne?%s&tag=%s" % (urlencode({'count': 1, 'display': "popular", 'size': "m", "layout":"x", "source":"all_tag"}), imagen))).read())
                try:
                    soaped = soup.findAll("div",id="flickr_badge_image1")[0].findAll("img")[0]["src"].replace("_m.jpg",".jpg")
                except IndexError:
                    soaped = "Lo siento, no puedo encontrar fotos de %s, si quieres busca tí en Google: " % imagen
                    try:
                        soaped+=" http://images.google.es/images?um=1&q="+imagen+"&btnG"
                    except:
                        exc_type, exc_value, exc_traceback = exc_info()
                        formatted_lines = format_exc().splitlines()
                        print " ".join(formatted_lines)
                varChat.privmsg( ch_active, "%s" % soaped)
                self.log+=ch_active+" "+strftime("(%Y-%m-%d %H:%M:%S) ",localtime(time()))+self.NickName+soaped+"\n"
            
    def on_nicknameinuse(self, varChat, varEvent):
        varChat.nick(varChat.get_nickname() + "_")

    def on_welcome(self, varChat, varEvent):
        varChat.join(self.channel)
        if  self.password != False:
            varChat.privmsg( "NickServ", "IDENTIFY %s %s" % (self.NickName, self.password))
        mens='he llegado para serviros'
        varChat.privmsg( self.channel, mens)
        for chann in self.channel.split(','):
            self.log+=chann+" "+strftime("(%Y-%m-%d %H:%M:%S) ",localtime(time()))+self.NickName+mens+"\n"

    def on_privmsg(self, varChat, varEvent):
        self.do_command(varEvent, varEvent.arguments()[0])
        #~ varSentence="".join(varEvent.arguments()[0].split(":", 1)).split()

    def on_pubmsg(self, varChat, varEvent):
        a = varEvent.arguments()[0].split(":", 1)
        #if len(a) > 1 and irc_lower(a[0]) == irc_lower(self.connection.get_nickname()):
            #self.do_command(varEvent, a[1].strip())
        self.do_command(varEvent, a[0].strip())
        return
        
    def show_status(self, varChat, varEvent):
        nick = nm_to_n(varEvent.source())
        for chname, chobj in self.channels.items():
            varChat.notice(nick, "--- Estadisticas del canal ---")
            varChat.notice(nick, "Channel: " + chname)
            users = chobj.users()
            users.sort()
            varChat.notice(nick, "Usuarios: " + ", ".join(users))
            opers = chobj.opers()
            opers.sort()
            varChat.notice(nick, "Operadores: " + ", ".join(opers))
            voiced = chobj.voiced()
            voiced.sort()
            varChat.notice(nick, "Voz: " + ", ".join(voiced))
            
    def on_dccmsg(self, varChat, varEvent):
        varChat.privmsg("Has dicho: " + varEvent.arguments()[0])
        
    def say_hello(self, varChat, varEvent):
        varChat.privmsg( varEvent.target(), "Muy buenas!")
        self.log+=varEvent.target()+" "+strftime("(%Y-%m-%d %H:%M:%S)",localtime(time()))+self.NickName+"Muy buenas!\n"
        
    def to_end_year(self, varChat, varEvent):
        ch_active = varEvent.target()
        now = date.today()
        newyear = date(2010, 12, 31)
        cik = newyear - now
        restdays = cik.days
        mens="Faltan %s días para fin de año" % str(restdays)
        varChat.privmsg( ch_active, mens )
        self.log+=ch_active+" "+strftime("(%Y-%m-%d %H:%M:%S)",localtime(time()))+self.NickName+mens+"\n"
        
    def show_rodadora(self, varChat, varEvent):
        ch_active = varEvent.target()
        varChat.privmsg( ch_active, "."*6+"@")
        varChat.privmsg( ch_active, "."*12+"@")
        varChat.privmsg( ch_active, "."*18+"@")
        varChat.privmsg( ch_active, "."*24+"@")
        active_time = strftime("(%Y-%m-%d %H:%M:%S) ",localtime(time()))
        self.log+=active_time+self.NickName+"."*6+"@"+"\n"+active_time+self.NickName+"."*12+"@"+"\n"+active_time+self.NickName+"."*18+"@"+"\n"+active_time+self.NickName+"."*24+"@\n"

    def on_dccchat(self, varChat, varEvent):
        if len(varEvent.arguments()) != 2:
            return
        args = varEvent.arguments()[1].split()
        if len(args) == 4:
            try:
                address = ip_numstr_to_quad(args[2])
                varPort = int(args[3])
            except ValueError:
                return
            self.dcc_connect(address, varPort)

    def do_command(self, varEvent, cmd):
        nick = nm_to_n(varEvent.source())
        varChat = self.connection
        ch_active = varEvent.target()
        varSentence="".join(varEvent.arguments()[0].split(":", 1)).split()
        varSay=nick+ ": "+" ".join(varSentence)

        timeline=ch_active+' '+strftime("(%Y-%m-%d %H:%M:%S) ",localtime(time()))+"".join(varSay)
        self.log+=timeline+"\n"
        print timeline
        #~ print len(varSentence)
        if varSentence[0][0] == '!':
            if self.flood.has_key(nick):
                sleep(1)
                self.flood[nick] = 0
            else:
                 self.flood[nick] = 1
            print self.flood
            if cmd == '!credits':
                self.show_credits(varChat, varEvent)
            if varSentence[0] == '!imagen':
                self.show_image(varChat, varEvent)
            #~ if 'mephiston' in varSentence:
                #~ varChat.privmsg( ch_active, 'Esta ausente, así que dudo que os conteste.')
            elif cmd == '!endyear':
                self.to_end_year(varChat, varEvent)
            elif cmd == '!rodadora':
                self.show_rodadora(varChat, varEvent)
            elif cmd == '!print':
                print self.log
                self.log=''
            else:
                varChat.notice(nick, "Not understood: " + cmd)
            

def main():
    import sys
    if len(sys.argv) not in [4,5]:
        print "Usage: testbot <server[:port]> <channel> <nickname>"
        sys.exit(1)

    varSystem = sys.argv[1].split(":", 1)
    varServer = varSystem[0]
    if len(varSystem) == 2:
        try:
            varPort = int(varSystem[1])
        except ValueError:
            print "Error: Puerto erroneo."
            sys.exit(1)
    else:
        varPort = 6667
    channel = sys.argv[2]
    varNickname = sys.argv[3]
    if len(sys.argv) == 5:
        varPassword = sys.argv[4]
        bot = TestBot(channel, varNickname, varPassword, varServer, varPort)
    else:
        varPassword=False
        bot = TestBot(channel, varNickname, varPassword, varServer, varPort)
    bot.start()

if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print "Interrumpido por teclado"


python-irclib/ircbot.py
# Copyright (C) 1999--2002  Joel Rosdahl
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA
#
# Joel Rosdahl <joel@rosdahl.net>
#
# $Id: ircbot.py,v 1.23 2008/09/11 07:38:30 keltus Exp $

"""ircbot -- Simple IRC bot library.

This module contains a single-server IRC bot class that can be used to
write simpler bots.
"""

import sys
from UserDict import UserDict
from irclib import SimpleIRCClient
from irclib import nm_to_n, irc_lower, all_events
from irclib import parse_channel_modes, is_channel
from irclib import ServerConnectionError

class SingleServerIRCBot(SimpleIRCClient):
    """A single-server IRC bot class.

    The bot tries to reconnect if it is disconnected.

    The bot keeps track of the channels it has joined, the other
    clients that are present in the channels and which of those that
    have operator or voice modes.  The "database" is kept in the
    self.channels attribute, which is an IRCDict of Channels.
    """
    def __init__(self, server_list, nickname, realname, reconnection_interval=60):
        """Constructor for SingleServerIRCBot objects.

        Arguments:

            server_list -- A list of tuples (server, port) that
                           defines which servers the bot should try to
                           connect to.

            nickname -- The bot's nickname.

            realname -- The bot's realname.

            reconnection_interval -- How long the bot should wait
                                     before trying to reconnect.

            dcc_connections -- A list of initiated/accepted DCC
            connections.
        """

        SimpleIRCClient.__init__(self)
        self.channels = IRCDict()
        self.server_list = server_list
        if not reconnection_interval or reconnection_interval < 0:
            reconnection_interval = 2**31
        self.reconnection_interval = reconnection_interval

        self._nickname = nickname
        self._realname = realname
        for i in ["disconnect", "join", "kick", "mode",
                  "namreply", "nick", "part", "quit"]:
            self.connection.add_global_handler(i,
                                               getattr(self, "_on_" + i),
                                               -10)

    def _connected_checker(self):
        """[Internal]"""
        if not self.connection.is_connected():
            self.connection.execute_delayed(self.reconnection_interval,
                                            self._connected_checker)
            self.jump_server()

    def _connect(self):
        """[Internal]"""
        password = None
        if len(self.server_list[0]) > 2:
            password = self.server_list[0][2]
        try:
            self.connect(self.server_list[0][0],
                         self.server_list[0][1],
                         self._nickname,
                         password,
                         ircname=self._realname)
        except ServerConnectionError:
            pass

    def _on_disconnect(self, c, e):
        """[Internal]"""
        self.channels = IRCDict()
        self.connection.execute_delayed(self.reconnection_interval,
                                        self._connected_checker)

    def _on_join(self, c, e):
        """[Internal]"""
        ch = e.target()
        nick = nm_to_n(e.source())
        if nick == c.get_nickname():
            self.channels[ch] = Channel()
        self.channels[ch].add_user(nick)

    def _on_kick(self, c, e):
        """[Internal]"""
        nick = e.arguments()[0]
        channel = e.target()

        if nick == c.get_nickname():
            del self.channels[channel]
        else:
            self.channels[channel].remove_user(nick)

    def _on_mode(self, c, e):
        """[Internal]"""
        modes = parse_channel_modes(" ".join(e.arguments()))
        t = e.target()
        if is_channel(t):
            ch = self.channels[t]
            for mode in modes:
                if mode[0] == "+":
                    f = ch.set_mode
                else:
                    f = ch.clear_mode
                f(mode[1], mode[2])
        else:
            # Mode on self... XXX
            pass

    def _on_namreply(self, c, e):
        """[Internal]"""

        # e.arguments()[0] == "@" for secret channels,
        #                     "*" for private channels,
        #                     "=" for others (public channels)
        # e.arguments()[1] == channel
        # e.arguments()[2] == nick list

        ch = e.arguments()[1]
        for nick in e.arguments()[2].split():
            if nick[0] == "@":
                nick = nick[1:]
                self.channels[ch].set_mode("o", nick)
            elif nick[0] == "+":
                nick = nick[1:]
                self.channels[ch].set_mode("v", nick)
            self.channels[ch].add_user(nick)

    def _on_nick(self, c, e):
        """[Internal]"""
        before = nm_to_n(e.source())
        after = e.target()
        for ch in self.channels.values():
            if ch.has_user(before):
                ch.change_nick(before, after)

    def _on_part(self, c, e):
        """[Internal]"""
        nick = nm_to_n(e.source())
        channel = e.target()

        if nick == c.get_nickname():
            del self.channels[channel]
        else:
            self.channels[channel].remove_user(nick)

    def _on_quit(self, c, e):
        """[Internal]"""
        nick = nm_to_n(e.source())
        for ch in self.channels.values():
            if ch.has_user(nick):
                ch.remove_user(nick)
    def die(self, msg="Adios mundo cruel"):
        """Let the bot die.

        Arguments:

            msg -- Quit message.
        """

        self.connection.disconnect(msg)
        sys.exit(0)

    def disconnect(self, msg="I'll be back!"):
        """Disconnect the bot.

        The bot will try to reconnect after a while.

        Arguments:

            msg -- Quit message.
        """
        self.connection.disconnect(msg)

    def get_version(self):
        """Returns the bot version.

        Used when answering a CTCP VERSION request.
        """
        return "pyBoter64 by Mephiston <meph.snake@gmail.com>"

    def jump_server(self, msg="Changing hosts"):
        """Connect to a new server, possibly disconnecting from the current.

        The bot will skip to next server in the server_list each time
        jump_server is called.
        """
        if self.connection.is_connected():
            self.connection.disconnect(msg)

        self.server_list.append(self.server_list.pop(0))
        self._connect()

    def on_ctcp(self, c, e):
        """Default handler for ctcp events.

        Replies to VERSION and PING requests and relays DCC requests
        to the on_dccchat method.
        """
        if e.arguments()[0] == "VERSION":
            c.ctcp_reply(nm_to_n(e.source()),
                         "VERSION " + self.get_version())
        elif e.arguments()[0] == "PING":
            if len(e.arguments()) > 1:
                c.ctcp_reply(nm_to_n(e.source()),
                             "PING " + e.arguments()[1])
        elif e.arguments()[0] == "DCC" and e.arguments()[1].split(" ", 1)[0] == "CHAT":
            self.on_dccchat(c, e)

    def on_dccchat(self, c, e):
        pass

    def start(self):
        """Start the bot."""
        self._connect()
        SimpleIRCClient.start(self)


class IRCDict:
    """A dictionary suitable for storing IRC-related things.

    Dictionary keys a and b are considered equal if and only if
    irc_lower(a) == irc_lower(b)

    Otherwise, it should behave exactly as a normal dictionary.
    """

    def __init__(self, dict=None):
        self.data = {}
        self.canon_keys = {}  # Canonical keys
        if dict is not None:
            self.update(dict)
    def __repr__(self):
        return repr(self.data)
    def __cmp__(self, dict):
        if isinstance(dict, IRCDict):
            return cmp(self.data, dict.data)
        else:
            return cmp(self.data, dict)
    def __len__(self):
        return len(self.data)
    def __getitem__(self, key):
        return self.data[self.canon_keys[irc_lower(key)]]
    def __setitem__(self, key, item):
        if key in self:
            del self[key]
        self.data[key] = item
        self.canon_keys[irc_lower(key)] = key
    def __delitem__(self, key):
        ck = irc_lower(key)
        del self.data[self.canon_keys[ck]]
        del self.canon_keys[ck]
    def __iter__(self):
        return iter(self.data)
    def __contains__(self, key):
        return self.has_key(key)
    def clear(self):
        self.data.clear()
        self.canon_keys.clear()
    def copy(self):
        if self.__class__ is UserDict:
            return UserDict(self.data)
        import copy
        return copy.copy(self)
    def keys(self):
        return self.data.keys()
    def items(self):
        return self.data.items()
    def values(self):
        return self.data.values()
    def has_key(self, key):
        return irc_lower(key) in self.canon_keys
    def update(self, dict):
        for k, v in dict.items():
            self.data[k] = v
    def get(self, key, failobj=None):
        return self.data.get(key, failobj)


class Channel:
    """A class for keeping information about an IRC channel.

    This class can be improved a lot.
    """

    def __init__(self):
        self.userdict = IRCDict()
        self.operdict = IRCDict()
        self.voiceddict = IRCDict()
        self.modes = {}

    def users(self):
        """Returns an unsorted list of the channel's users."""
        return self.userdict.keys()

    def opers(self):
        """Returns an unsorted list of the channel's operators."""
        return self.operdict.keys()

    def voiced(self):
        """Returns an unsorted list of the persons that have voice
        mode set in the channel."""
        return self.voiceddict.keys()

    def has_user(self, nick):
        """Check whether the channel has a user."""
        return nick in self.userdict

    def is_oper(self, nick):
        """Check whether a user has operator status in the channel."""
        return nick in self.operdict

    def is_voiced(self, nick):
        """Check whether a user has voice mode set in the channel."""
        return nick in self.voiceddict

    def add_user(self, nick):
        self.userdict[nick] = 1

    def remove_user(self, nick):
        for d in self.userdict, self.operdict, self.voiceddict:
            if nick in d:
                del d[nick]

    def change_nick(self, before, after):
        self.userdict[after] = 1
        del self.userdict[before]
        if before in self.operdict:
            self.operdict[after] = 1
            del self.operdict[before]
        if before in self.voiceddict:
            self.voiceddict[after] = 1
            del self.voiceddict[before]

    def set_mode(self, mode, value=None):
        """Set mode on the channel.

        Arguments:

            mode -- The mode (a single-character string).

            value -- Value
        """
        if mode == "o":
            self.operdict[value] = 1
        elif mode == "v":
            self.voiceddict[value] = 1
        else:
            self.modes[mode] = value

    def clear_mode(self, mode, value=None):
        """Clear mode on the channel.

        Arguments:

            mode -- The mode (a single-character string).

            value -- Value
        """
        try:
            if mode == "o":
                del self.operdict[value]
            elif mode == "v":
                del self.voiceddict[value]
            else:
                del self.modes[mode]
        except KeyError:
            pass

    def has_mode(self, mode):
        return mode in self.modes

    def is_moderated(self):
        return self.has_mode("m")

    def is_secret(self):
        return self.has_mode("s")

    def is_protected(self):
        return self.has_mode("p")

    def has_topic_lock(self):
        return self.has_mode("t")

    def is_invite_only(self):
        return self.has_mode("i")

    def has_allow_external_messages(self):
        return self.has_mode("n")

    def has_limit(self):
        return self.has_mode("l")

    def limit(self):
        if self.has_limit():
            return self.modes["l"]
        else:
            return None

    def has_key(self):
        return self.has_mode("k")

    def key(self):
        if self.has_key():
            return self.modes["k"]
        else:
            return None
